pkg/sqlite/load.go

```diff
@@ -113,39 +113,140 @@ func (s *SQLLoader) AddPackageChannels(manifest registry.PackageManifest) error
 		tx.Rollback()
 	}()
 
-	addPackage, err := tx.Prepare("insert into package(name) values(?)")
-	if err != nil {
-		return err
+	var errs []error
+
+	if err := addPackageIfNotExists(tx, graph.Name); err != nil {
+		errs = append(errs, err)
 	}
-	defer addPackage.Close()
 
-	addDefaultChannel, err := tx.Prepare("update package set default_channel = ? where name = ?")
-	if err != nil {
-		return err
+	for name, channel := range graph.Channels {
+		if err := addOrUpdateChannel(tx, name, graph.Name, channel.Head.CsvName); err != nil {
+			errs = append(errs, err)
+			continue
+		}
 	}
-	defer addDefaultChannel.Close()
 
-	addChannel, err := tx.Prepare("insert into channel(name, package_name, head_operatorbundle_name) values(?, ?, ?)")
-	if err != nil {
-		return err
+	if err := updateDefaultChannel(tx, graph.DefaultChannel, graph.Name); err != nil {
+		errs = append(errs, err)
 	}
-	defer addChannel.Close()
 
-	addChannelEntry, err := tx.Prepare("insert into channel_entry(channel_name, package_name, operatorbundle_name, depth) values(?, ?, ?, ?)")
+	// update each channel's graph
+	for name, channel := range graph.Channels {
+		currentNode := channel.Head
+		depth := 1
+
+		var previousNodeID int64
+
+		// first clear the current channel graph
+		err := truncChannelGraph(tx, name, graph.Name)
+		if err != nil {
+			errs = append(errs, err)
+			break
+		}
+
+		// iterate into the replacement chain of the channel to insert or update all entries
+		for {
+			// create real channel entry for node
+			id, err := addChannelEntry(tx, name, graph.Name, currentNode.CsvName, depth)
+			if err != nil {
+				errs = append(errs, err)
+				break
+			}
+
+			channelEntryCSV, err := s.getCSV(tx, currentNode.CsvName)
+			if err != nil {
+				errs = append(errs, err)
+				break
+			}
+
+			if err := s.addAPIs(tx, channelEntryCSV, id); err != nil {
+				errs = append(errs, err)
+			}
+
+			// If the previous node was created, use the entryId of the current node to update
+			// the replaces for the previous node
+			if previousNodeID != 0 {
+				err := addReplaces(tx, id, previousNodeID)
+				if err != nil {
+					errs = append(errs, err)
+				}
+			}
+
+			syntheticReplaces := make([]registry.BundleKey, 0) // for CSV skips
+			nextNode := registry.BundleKey{}
+
+			// Iterate over all replaces for the node in the graph
+			// It should only contain one real replacement, so let's find it and
+			// follow the chain. For the rest, they are fake entries and should be
+			// generated as synthetic replacements
+			for replace := range channel.Nodes[currentNode] {
+				if _, ok := channel.Nodes[replace]; !ok {
+					syntheticReplaces = append(syntheticReplaces, replace)
+				} else {
+					nextNode = replace
+				}
+			}
```

Comment:
we should catch when the graph isn't what we expect here - where we don't find any "real" replace in the graph

Reply:
Is there a difference between "no real" replaces and "no real replaces but has some synthetic replaces"? What about the case where there is no replacement but the first node in the graph has skips defined? Are we saying that's not valid?

Reply:
It looks to me like there is a difference -

We expect `channel.Nodes[replaces]` to exist, but not `channel.Nodes[skips]`

This block will find the `replaces` node if it exists, and add all the `skip` nodes if they exist - which is good. But if the `replaces` node doesn't exist, this will set `replaces` to whatever happens to be the last `skips` node, which would be incorrect (especially if elsewhere we assume `replaces` exists in `channel.Nodes`)

Reply:
I don't think it will. If the replacement node isn't in the graph, it will just never set `nextNode` at all and assume the graph ended (since just before this I initialized `nextNode` as an empty node). In semver insert mode, this implies that either we really are at the last node OR we aren't at the last node and the real replacement wasn't defined properly in the graph. In that case, we can check that the actual replacement chain is as long as the set of real nodes defined in the graph. If they aren't, it means we got to the end early.

But, when I was thinking about this I realized that this code is non-deterministic for graphs built with `skippatch` mode -- in that mode there are multiple real candidates for `nextNode` and the graph isn't being inserted properly. Right now this is definitely broken. So I also need to update the `nextNode` to a list of nodes to iterate over and insert